Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,322 @@
+            })
+        }
+
+        internal func cancelSystemMetrics() {
+            self.task?.cancel()
+            self.task = nil
+        }
+
+        internal func poll() {
+            self.queue.asyncAfter(deadline: .now() + self.timeInterval) {
+                guard let task = self.task else { return }
+                task.perform()
+                self.poll()
+            }
+        }
+        
+        func makeCounter(label: String, dimensions: [(String, String)]) -> CounterHandler {
+            self.underlying.makeCounter(label: label, dimensions: dimensions)
+        }
+        
+        func makeRecorder(label: String, dimensions: [(String, String)], aggregate: Bool) -> RecorderHandler {
+            self.underlying.makeRecorder(label: label, dimensions: dimensions, aggregate: aggregate)
+        }
+        
+        func makeTimer(label: String, dimensions: [(String, String)]) -> TimerHandler {
+            self.underlying.makeTimer(label: label, dimensions: dimensions)
+        }
+        
+        func destroyCounter(_ handler: CounterHandler) {
+            self.underlying.destroyCounter(handler)
+        }
+        
+        func destroyRecorder(_ handler: RecorderHandler) {
+            self.underlying.destroyRecorder(handler)
+        }
+        
+        func destroyTimer(_ handler: TimerHandler) {
+            self.underlying.destroyTimer(handler)
+        }
+    }
+}
+
+public protocol SystemMetricsProvider {
+    static func readSystemMetrics() -> SystemMetrics.Data?
+}
+
+public enum SystemMetrics {
+    /// Options used to bootstrap `SystemMetricsHandler`
+    ///
+    /// Libraries are advised to extend `SystemMetricsOptions` with a static instance for ease of use.
+    public struct Options {
+        let interval: DispatchTimeInterval
+        let metricsType: SystemMetricsProvider.Type?
+        let labels: Labels
+        
+        /// Create new instance of `SystemMetricsOptions`
+        ///
+        /// - parameters:
+        ///     - pollInterval: The interval at which system metrics should be updated.
+        ///     - systemMetricsType: The type of system metrics to use. If none is provided this defaults to
+        ///                          `LinuxSystemMetrics` on Linux platforms and `NOOPSystemMetrics` on all other platforms.
+        ///     - labels: The labels to use for generated system metrics.
+        public init(pollInterval interval: DispatchTimeInterval = .seconds(2), metricsType: SystemMetricsProvider.Type? = nil, labels: Labels) {
+            self.interval = interval
+            self.metricsType = metricsType
+            self.labels = labels
+        }
+    }
+
+    /// Labels for the reported System Metrics.
+    ///
+    /// Backend implementations are encouraged to provide a static extension to this type with
+    /// defaults that suit their specific backend needs.
+    public struct Labels {
```

Comment:
do we want to provide default labels?

Reply:
My idea here was to have implementations provide default labels for their backend/style choices. They could then also add default overloads for all the other values.

Personally don't think we should ship labels with this package since it's a very backend specific detail

Reply:
@ktoso wdyt?